File: leetcode/LinkedListTests/LinkedList_206.cs
namespace LinkedListTests;

[TestFixture]
class LinkedList_206
{
    private class SolutionRecursion {
        public ListNode? ReverseList(ListNode? head) {
            if(head is null || head.next is null) { 
                return head; 
            }
            var last = ReverseList(head.next);
            head.next.next = head;
            head.next = null;
            return last;
        }
    }
    
    private class SolutionIterative {
        public ListNode? ReverseList(ListNode? head)
        {
            ListNode? prev = null;
            var curr = head;
            while (curr is not null)
            {
                var nextTmp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = nextTmp;
            }
            return prev;
        }
    }
}

File: leetcodePython/LinkedList/linkedlist_206.py
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # prev -> curr -> next
        # next -> curr -> prev

        # none -> head -> next
        # none <- head <- next

        prev = None
        curr = head
        while curr:
            curr_next = curr.next
            # none <- head  or prev <- curr
            curr.next = prev
            prev = curr
            curr = curr_next
        return prev
